File: src/commands/git/worktree.ts
import path from "node:path";
import { intro, log, outro, spinner } from "@clack/prompts";
import { Command } from "commander";
import { getShellError } from "../../utils/errors";

type Shell = typeof Bun.$;

export async function createWorktree(
	shell: Shell,
	name: string,
	opts: { base: string },
): Promise<{ worktreePath: string }> {
	const repoRoot = await shell`git rev-parse --show-toplevel`.quiet().text();
	const worktreePath = path.join(path.dirname(repoRoot.trim()), name);
	await shell`git fetch origin ${opts.base}`.quiet();
	await shell`git worktree add -b ${name} ${worktreePath} origin/${opts.base}`.quiet();
	return { worktreePath };
}

export const worktree = new Command("worktree")
	.description("Create a new worktree from a base branch")
	.argument("<name>", "Name for the new branch and worktree directory")
	.option("-b, --base <branch>", "Base branch to create from", "master")
	.action(async (name: string, opts: { base: string }) => {
		intro("git worktree");

		const s = spinner();
		try {
			const repoRoot = (
				await Bun.$`git rev-parse --show-toplevel`.quiet().text()
			).trim();
			const worktreePath = path.join(path.dirname(repoRoot), name);

			s.start(`Fetching origin/${opts.base}...`);
			await Bun.$`git fetch origin ${opts.base}`.quiet();
			s.stop("Fetched");

			s.start(`Creating worktree at ${worktreePath}...`);
			await Bun.$`git worktree add -b ${name} ${worktreePath} origin/${opts.base}`.quiet();
			s.stop("Created");

			outro(`Worktree ready at ${worktreePath}`);
		} catch (e: unknown) {
			s.stop("Failed");
			log.error(getShellError(e));
			process.exit(1);
		}
	});

File: src/commands/git/worktree.rs
use crate::utils::git::Shell;
use anyhow::Result;

fn run_cmd(args: &[&str]) -> Result<String> {
    let output = std::process::Command::new("git").args(args).output()?;
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow::anyhow!("git failed: {stderr}"));
    }
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

pub fn create_worktree(shell: &dyn Shell, name: &str, base: &str) -> Result<String> {
    let root = shell.run("git", &["rev-parse", "--show-toplevel"])?;
    let trimmed = root.trim().to_string();
    let parent = std::path::Path::new(&trimmed)
        .parent()
        .unwrap_or(std::path::Path::new("."));
    let worktree_path = parent.join(name);

    run_cmd(&["fetch", "origin", base])?;
    run_cmd(&[
        "worktree",
        "add",
        "-b",
        name,
        worktree_path.to_str().unwrap(),
        &format!("origin/{base}"),
    ])?;

    Ok(worktree_path.to_str().unwrap().to_string())
}

pub fn execute_worktree(shell: &dyn Shell, name: &str, base: &str) -> Result<()> {
    let root = shell.run("git", &["rev-parse", "--show-toplevel"])?;
    let trimmed = root.trim().to_string();
    let parent = std::path::Path::new(&trimmed)
        .parent()
        .unwrap_or(std::path::Path::new("."));
    let worktree_path = parent.join(name);

    println!("Fetching origin/{base}...");
    run_cmd(&["fetch", "origin", base])?;
    println!("Fetched");

    println!("Creating worktree at {}...", worktree_path.display());
    run_cmd(&[
        "worktree",
        "add",
        "-b",
        name,
        worktree_path.to_str().unwrap(),
        &format!("origin/{base}"),
    ])?;
    println!("Created");

    println!("Worktree ready at {}", worktree_path.display());
    Ok(())
}

#[cfg(test)]
mod tests {
    use crate::utils::git::Shell;
    use std::sync::Mutex;

    struct MockShell {
        responses: Mutex<Vec<anyhow::Result<String>>>,
    }

    impl MockShell {
        fn new(responses: Vec<anyhow::Result<String>>) -> Self {
            Self {
                responses: Mutex::new(responses),
            }
        }
    }

    impl Shell for MockShell {
        fn run(&self, _cmd: &str, _args: &[&str]) -> anyhow::Result<String> {
            self.responses.lock().unwrap().remove(0)
        }
    }

    #[test]
    fn test_create_worktree_resolves_path() {
        let shell = MockShell::new(vec![Ok("/home/user/repo".to_string())]);
        let result = shell.run("git", &["rev-parse", "--show-toplevel"]);
        assert_eq!(result.unwrap(), "/home/user/repo");
    }
}
